soc/arm/nxp_kinetis/k6x/power.c

```diff
@@ -0,0 +1,153 @@
+
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+void enable_uart(void);
+void disable_uart(void);
+
+#if defined(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME)
+static ALWAYS_INLINE void set_clock_hw_cycles_per_sec(int cycles)
+{
+	extern int z_clock_hw_cycles_per_sec;
+
+	z_clock_hw_cycles_per_sec = cycles;
+}
+#else
+static ALWAYS_INLINE void set_clock_hw_cycles_per_sec(int cycles) {}
+#endif
+
+static ALWAYS_INLINE void set_clock_to_FEI_mode(void)
+{
+	const mcg_config_t mcgConfigStruct = {
+		/* FEI - FLL with internal RTC. */
+		.mcgMode = kMCG_ModeFEI,
+		/* MCGIRCLK enabled, MCGIRCLK disabled in STOP mode */
+		.irclkEnableMode = kMCG_IrclkEnable,
+		/* Slow internal reference clock selected */
+		.ircs = kMCG_IrcSlow,
+		/* Fast IRC divider: divided by 1 */
+		.fcrdiv = 0x0U,
+		/* FLL reference clock divider: divided by 32 */
+		.frdiv = 0x0U,
+		/* Low frequency range */
+		.drs = kMCG_DrsLow,
+		/* DCO has a default range of 25% */
+		.dmx32 = kMCG_Dmx32Default,
+		/* Selects System Oscillator (OSCCLK) */
+		.oscsel = kMCG_OscselOsc,
+		.pll0Config = {
+			/* MCGPLLCLK disabled */
+			.enableMode = 0,
+			/* PLL Reference divider: divided by 20 */
+			.prdiv = 0x13U,
+			/* VCO divider: multiplied by 48 */
+			.vdiv = 0x18U,
+		},
+	};
+
+	CLOCK_SetMcgConfig(&mcgConfigStruct);
+
+	set_clock_hw_cycles_per_sec(25000000);
+}
+
+static ALWAYS_INLINE void set_clock_to_PEE_mode(void)
+{
+	const mcg_config_t mcgConfigStruct = {
+		/* PEE - PLL Engaged External */
+		.mcgMode = kMCG_ModePEE,
+		/* MCGIRCLK enabled, MCGIRCLK disabled in STOP mode */
+		.irclkEnableMode = kMCG_IrclkEnable,
+		/* Slow internal reference clock selected */
+		.ircs = kMCG_IrcSlow,
+		/* Fast IRC divider: divided by 1 */
+		.fcrdiv = 0x0U,
+		/* FLL reference clock divider: divided by 32 */
+		.frdiv = 0x0U,
+		/* Low frequency range */
+		.drs = kMCG_DrsLow,
+		/* DCO has a default range of 25% */
+		.dmx32 = kMCG_Dmx32Default,
+		/* Selects System Oscillator (OSCCLK) */
+		.oscsel = kMCG_OscselOsc,
+		.pll0Config = {
+			/* MCGPLLCLK disabled */
+			.enableMode = 0,
+			/* PLL Reference divider: divided by 20 */
+			.prdiv = 0x13U,
+			/* VCO divider: multiplied by 48 */
+			.vdiv = 0x18U,
+		},
+	};
+
+	CLOCK_SetMcgConfig(&mcgConfigStruct);
+
+	set_clock_hw_cycles_per_sec(120000000);
+}
+
+static ALWAYS_INLINE void enter_sleep_mode_1(void)
+{
+	disable_uart();
+	set_clock_to_FEI_mode();
+	SMC_PreEnterWaitModes();
+	SMC_SetPowerModeWait(SMC);
+	SMC_PostExitWaitModes();
+}
+
+static ALWAYS_INLINE void exit_sleep_mode_1(void)
+{
+	enable_uart();
+	set_clock_to_PEE_mode();
+}
+#endif
+
+void sys_set_power_state(enum power_states state)
+{
+	__disable_irq();
+	irq_unlock(0);
+
+	switch (state) {
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+#ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_1
+	case SYS_POWER_STATE_SLEEP_1:
+	{
+		enter_sleep_mode_1();
+		break;
+	}
+#endif
+#endif
+	default:
+		LOG_INF("Unsupported power state %u", state);
+		break;
+	}
+	__enable_irq();
+}
```

Comment:
This should be moved to the end of _sys_pm_power_state_exit_post_ops. Other power management drivers are using irq_unlock(0) to reenable, I would suggest doing the same.

Reply:
I will have a look at it.